### DISCUSSION TITLE
jumping to line number in editor

### DISCUSSION THREAD

**Discussion by khimaros** [zk-org/zk#330] [Ideas]
### Verify

- [X] I searched the existing discussions for similar ideas

### Share your idea or feature request

it would be awesome to jump to the line number of where text was found

for example, `zk edit -i`, search for a string, press enter, vim opens at that line.

vim itself supports a `+<lnum>` eg. `+23` to jump to line 23.

one way to support this generically would be to have var expansion `{{line-number}}`

then we could use this in the `editor` field `editor = "vim +{{line-number}}"`

**Comment by mickael-menu** [comment]
The `{{line-number}}` expansion is a great idea for implementing this, and maybe it could be done for stuff like `zk list --match ...`. But for `zk edit -i`, I'm not sure how that would work. Is there a way for `fzf` to return the line number of the match?

**Comment by khimaros** [comment]
here is the recommended way to do this with fzf: https://github.com/junegunn/fzf/issues/1014#issuecomment-321974618

the `nl` tool also has a `-s` switch which can be used to set the delimiter between the line number and the line content.

**Reply by mickael-menu** [reply]
This returns the index of the selection list, which is the list of notes and not their content.

**Reply by khimaros** [reply]
@mickael-menu, i think it would be necessary to make a second invocation to fzf (eg. choosing best match) after the file has been selected in interactive mode.

for posterity, here's my understanding of how zk works today:

- the contents of each file in the database is compressed down to a single line (line breaks removed) with the full path to the file appended to the end
- a list of those file content + file path lines are passed to fzf, one file per line
- fzf returns the file content + file path pair, and zk parses that to extract the full file path and then opens that file with the configured editor

**Reply by mickael-menu** [reply]
Yes that's how it is implemented in `zk` currently. Except for third step, we can ask `fzf` to return a particular field in the line (there are invisible characters separating the "columns"), so we get only the file path and not the content.

I believe you could work out a `zk alias` to start a second instance of `fzf` to choose the line to open.

**Reply by khimaros** [reply]
@mickael-menu i'd love to do this but haven't been able to figure out the right incantation. is there a chance you could guide me in the right direction?

**Reply by khimaros** [reply]
here's what i have working on the shell:

```shell
FILE=some.md
grep -nEv '^[[:space:]]*$' "$FILE" \
    | fzf \
        --tiebreak=begin \
        --exact \
        --tabstop=4 \
        --height=100% \
        --layout=reverse \
        --no-hscroll \
        --color=hl:-1,hl+:-1 \
        --preview-window=wrap \
        --delimiter=':' \
        --with-nth=2.. \
    | sed 's/:.*$//' \
    | xargs -o -I {} vim +{} "$FILE"
```

all fzf flags except `--delimiter=` and `--with-nth=` are from zk `defaultOptions`

what i'm not clear on is how to integrate this cleanly with the zk editor and other commands.

**Comment by khimaros** [comment]
okay, the solution i've settled on is to have a bash script in my path named `vf` which contains:

```bash
#!/bin/bash

set -eou pipefail

grep -nEv '^[[:space:]]*$' "$1" \
	| fzf \
		--tiebreak=begin \
		--exact --tabstop=4 \
		--height=100% \
		--layout=reverse \
		--no-hscroll \
		--color=hl:-1,hl+:-1 \
		--preview-window=wrap \
		--delimiter=':' \
		--with-nth=2.. \
	| sed 's/:.*$//' \
	| xargs -o -I {} vim +{} "$1"
```

and then set in `.zk/config.toml`

```toml
[tools]
editor = 'vf'
```

**Reply by khimaros** [reply]
i think the missing piece here is passing the query from the first fzf invocation to the second one.

is there some template or environment variable that could be used for that?

**Reply by khimaros** [reply]
early draft https://github.com/mickael-menu/zk/compare/main...khimaros:zk:main which populates a `$ZK_QUERY` env var when invoking the editor.

this means the `vf` command above can pass `--query $ZK_QUERY` to fzf

it's possible this patch breaks some other behavior of zk.

@mickael-menu curious about your thoughts on this approach.

**Reply by khimaros** [reply]
now tracking at https://github.com/mickael-menu/zk/pull/361

**Comment by khimaros** [comment]
FYI, i am maintaining a fork with this functionality since upstream is not interested in accepting a PR: https://github.com/khimaros/zk/
